src/stores/widgets/StopGapWidgetDriver.ts

```diff
@@ -328,6 +333,93 @@ export class StopGapWidgetDriver extends WidgetDriver {
         return { roomId, eventId: r.event_id };
     }
 
+    /**
+     * @experimental Part of MSC4140 & MSC4157
+     */
+    public async sendDelayedEvent<K extends keyof StateEvents>(
+        delay: number | null,
+        parentDelayId: string | null,
+        eventType: K,
+        content: StateEvents[K],
+        stateKey: string | null,
+        targetRoomId: string | null,
+    ): Promise<ISendDelayedEventDetails>;
+    /**
+     * @experimental Part of MSC4140 & MSC4157
+     */
+    public async sendDelayedEvent<K extends keyof TimelineEvents>(
+        delay: number | null,
+        parentDelayId: string | null,
+        eventType: K,
+        content: TimelineEvents[K],
+        stateKey: null,
+        targetRoomId: string | null,
+    ): Promise<ISendDelayedEventDetails>;
+    public async sendDelayedEvent(
+        delay: number | null,
+        parentDelayId: string | null,
+        eventType: string,
+        content: IContent,
+        stateKey: string | null = null,
+        targetRoomId: string | null = null,
+    ): Promise<ISendDelayedEventDetails> {
+        const client = MatrixClientPeg.get();
+        const roomId = targetRoomId || SdkContextClass.instance.roomViewStore.getRoomId();
+
+        if (!client || !roomId) throw new Error("Not in a room or not attached to a client");
+
+        let delayOpts;
+        if (delay !== null) {
+            delayOpts = {
+                delay,
+                ...(parentDelayId !== null && { parent_delay_id: parentDelayId }),
+            };
+        } else if (parentDelayId !== null) {
+            delayOpts = {
+                parent_delay_id: parentDelayId,
+            };
+        } else {
+            throw new Error("Must provide at least one of delay or parentDelayId");
```

Comment:
i think we should use the `parent_delay_id` wording here since this is an error returned to the widget and the widget will not use the camel case format.

Reply:
This is a tricky one...using snake case implies that what's missing here is a field of the `send_event` widget, but what's actually missing is a parameter of an internal call within the embedded client.

If this error is hit, it means the client side of the widget-api didn't check a received `send_event` payload properly, and asked the driver to send a delayed event with no delay-related parameters, when it should have instead asked to send a non-delayed event.

In practice, this will never happen, since the widget-api indeed does all that checking, but there's no way for driver implementations to know that.  The best way to avoid all this would be to use better types on the parameter list of `sendDelayedEvent`, so that it's obvious from the driver's perspective that either one or both of the delay-related parameters will be present.

...but that's a lot of faff for something that might not even become stable, so for now I'll just leave this as-is. :slightly_smiling_face:

Reply:
yes leave it as is. I actually was not fully aware, that this just for the driver part.